conseil-lorre/src/main/scala/tech/cryptonomic/conseil/indexer/tezos/TezosDatabaseOperations.scala

```diff
@@ -890,4 +892,290 @@ object TezosDatabaseOperations extends LazyLogging {
         Future.successful(List.empty -> None)
     }
 
+  /** Write an audit log entry of a detected fork and some
+    * reference data useful to analyse the event.
+    *
+    * @param forkLevel the point where data was actually different, the fork point
+    * @param forkHash the first block hash differing from that on the current chain
+    * @param indexedHeadLevel how far has the indexer gone ahead in the fork
+    * @param detectionTime when the indexer identified the fork having happened
+    * @return a unique identifier for the newly recorded fork, used as `forkId` in the system
+    */
+  def writeForkEntry(
+      forkLevel: BlockLevel,
+      forkHash: TezosBlockHash,
+      indexedHeadLevel: BlockLevel,
+      detectionTime: Instant
+  ): DBIO[String] = {
+    val forkId = ju.UUID.randomUUID().toString
+    val ts = new Timestamp(detectionTime.getEpochSecond())
+    Tables.Forks.returning(
+      Tables.Forks.map(_.forkId)
+    ) += ForksRow(forkId, forkLevel, forkHash.value, indexedHeadLevel, ts)
+  }
+
+  /** Temporarily lift statement constraints on foreign keys
+    * We need to defer any such contraints until the transaction commits
+    * when we want to update fork references in blocks and all related
+    * db entities, lifting the constraint checks until everything is updated
+    * and consistent.
+    */
+  def deferForkConstraints(): DBIO[Int] =
+    sqlu"SET CONSTRAINTS ALL DEFERRED;"
+
+  /** Marks all relevant entities as invalidated (i.e. by a forking event), by
+    * specifying the block level at which the chain showed divergence from the local data.
+    *
+    * Such invalidated entries should not appear anymore as results from queries against the
+    * main fork of the chain. They will thereafter need to be specifically requested.
+    *
+    * @param fromLevel the lower level, included, for which data doesn't match the node
+    * @param asOf a time-stamp of the invalidation operation
+    * @param forkId a unique identifier of the fork on which the entities were found
+    * @return the number of impacted rows
+    */
+  def invalidateBlocks(fromLevel: BlockLevel, asOf: Instant, forkId: String): DBIO[Int] = {
```

Comment:
Did you investigate possibility of making one generic invalidation method? it looks like a lot of copy-pasted code here.

Reply:
I'll check it out, thanks

Reply:
I was honestly unconvinced that the result would be actually usable, yet I'm surprised with the result.
I think I found a way to make it easy enough to follow the logic and keep it simple at call site too.
Let me know what you think

Reply:
Definitely looks, better. Great job!